File: program_code/local_model_tools/cost_gate.py
"""
Cost-Aware Entry Gate — Pre-trade volatility vs. cost check
成本感知入場門檻 — 開倉前波動率 vs 交易成本檢查

MODULE_NOTE (中文):
  本模組在開倉前檢查預期波動是否足以覆蓋交易成本（手續費 + 滑點）。
  背景：Demo 交易數據顯示 14 筆贏利交易中 11 筆 gross profit < 手續費，
  即「贏了但實際虧錢」。此門檻確保只在波動足夠時才開倉。

  設計原則：
  - 確定性規則，不依賴歷史交易表現
  - Fail-open：ATR 數據不可用時放行（不阻塞冷啟動）
  - Per-symbol 成本計算（不同幣種手續費 + 滑點不同）
  - 安全閥：不能造成看盤一天完全沒有成交（Operator 2026-04-02 決策）

  公式：
    c_round_pct = (taker_fee + slippage) × 2          # 開+平倉各一次
    min_expected_move_pct = c_round_pct / max(0.3, win_rate) × 1.3  # 安全邊際
    若 ATR% < min_expected_move_pct → 拒絕開倉

MODULE_NOTE (English):
  This module checks whether expected volatility covers trading costs before opening.
  Background: Demo data showed 11 of 14 winning trades had gross profit < fees,
  meaning "wins that actually lost money". This gate ensures sufficient volatility.

  Design principles:
  - Deterministic rule, no dependency on historical trade performance
  - Fail-open: passes through when ATR data unavailable (no cold-start blocking)
  - Per-symbol cost calculation (different symbols have different fee + slippage)
  - Safety valve: must not cause zero trades for a full day (Operator decision 2026-04-02)

  Formula:
    c_round_pct = (taker_fee + slippage) × 2
    min_expected_move_pct = c_round_pct / max(0.3, win_rate) × 1.3
    If ATR% < min_expected_move_pct → reject entry

Safety invariant / 安全不變量:
  - 純決策邏輯，不執行任何交易操作 / Pure decision logic, no trading operations
  - Fail-open on missing data / 數據缺失時放行
  - 原則 5（生存 > 利潤）+ 原則 13（成本感知）
"""

from __future__ import annotations

import logging

logger = logging.getLogger(__name__)


# ─── Fee Constants / 手續費常量 ───
# Bybit linear USDT perpetual taker fee rate
# Bybit 永續合約 taker 手續費率
BYBIT_TAKER_FEE_RATE: float = 0.00055  # 0.055%


# ─── Slippage Tiers / 滑點分級 ───
# Source: paper_trading_engine.py SLIPPAGE_TIERS (duplicated to avoid circular import)
# 來源：paper_trading_engine.py SLIPPAGE_TIERS（複製以避免循環依賴）
# Sorted descending by 24h USD turnover threshold.
# 按 24h 成交額閾值降序排列。
SLIPPAGE_TIERS: list[tuple[float, float]] = [
    (1_000_000_000, 0.0001),   # >$1B turnover: 1 bps (BTC/ETH)
    (100_000_000,   0.0002),   # >$100M: 2 bps
    (10_000_000,    0.0005),   # >$10M: 5 bps
    (1_000_000,     0.0015),   # >$1M: 15 bps
    (0,             0.0030),   # <$1M: 30 bps (illiquid alts)
]

# Default slippage when volume data is unavailable / 無成交量數據時的默認滑點
DEFAULT_SLIPPAGE_RATE: float = 0.0005  # 5 bps


def _lookup_slippage(volume_24h: float) -> float:
    """
    Look up slippage rate by 24h volume tier.
    根據 24h 成交量查找滑點率。

    Args:
        volume_24h: 24-hour trading volume in USD / 24 小時成交量（USD）

    Returns:
        Slippage rate as a decimal (e.g. 0.0005 = 5 bps) / 滑點率（小數）
    """
    if volume_24h <= 0:
        return DEFAULT_SLIPPAGE_RATE
    for threshold, rate in SLIPPAGE_TIERS:
        if volume_24h >= threshold:
            return rate
    return DEFAULT_SLIPPAGE_RATE


def compute_round_trip_cost_pct(
    symbol: str,
    volume_24h: float = 0.0,
    taker_fee_rate: float = BYBIT_TAKER_FEE_RATE,
) -> float:
    """
    Compute round-trip cost as a percentage of notional value.
    計算來回交易成本佔名義價值的百分比。

    Round trip = open + close, each incurs taker fee + slippage.
    一來一回 = 開倉 + 平倉，各產生 taker 手續費 + 滑點。

    Args:
        symbol:         Trading pair (for logging) / 交易對（用於日誌）
        volume_24h:     24h volume in USD for slippage lookup / 24h 成交量
        taker_fee_rate: Taker fee rate (default Bybit linear) / Taker 手續費率

    Returns:
        Round-trip cost as percentage (e.g. 0.21 means 0.21%)
        來回成本百分比（如 0.21 表示 0.21%）
    """
    slippage_rate = _lookup_slippage(volume_24h)
    # Each leg: taker_fee + slippage; two legs for round trip
    # 每一腿：手續費 + 滑點；來回共兩腿
    cost_pct = (taker_fee_rate + slippage_rate) * 2 * 100.0  # convert to percentage
    return cost_pct


def should_reject_for_cost(
    symbol: str,
    atr_pct: float | None,
    win_rate: float = 0.5,
    daily_trade_count: int = 0,
    volume_24h: float = 0.0,
) -> tuple[bool, str]:
    """
    Decide whether to reject a trade entry due to insufficient volatility.
    決定是否因波動率不足拒絕開倉。

    Logic:
      1. ATR unavailable → fail-open (pass through, principle 6: uncertain → conservative
         but don't block cold-start)
      2. Compute min_expected_move_pct = c_round / max(0.3, win_rate) * 1.3
      3. Safety valve: if no trades today and ATR > half of round-trip cost → allow
         (Operator decision: avoid zero-trade days)
      4. If ATR% < min_expected_move_pct → reject

    邏輯：
      1. ATR 不可用 → fail-open（放行，原則 6：不確定時保守但不阻塞冷啟動）
      2. 計算 min_expected_move_pct = 來回成本 / max(0.3, 勝率) × 1.3
      3. 安全閥：若今天尚無成交且 ATR > 來回成本一半 → 放行
         （Operator 決策：避免整天零成交）
      4. 若 ATR% < min_expected_move_pct → 拒絕

    Args:
        symbol:            Trading pair / 交易對
        atr_pct:           ATR as % of price (e.g. 1.5 = 1.5%), or None / ATR 佔價格百分比
        win_rate:          Estimated win rate (0.0-1.0, default 0.5) / 估計勝率
        daily_trade_count: Number of trades already executed today / 今天已成交數
        volume_24h:        24h volume in USD for slippage lookup / 24h 成交量

    Returns:
        (should_reject: bool, reason: str)
        (是否拒絕, 原因字符串)
    """
    # ── Fail-open: no ATR data → allow (cold-start safety) ──
    # 無 ATR 數據 → 放行（冷啟動安全）
    if atr_pct is None:
        return (False, "atr_unavailable_pass_through")

    # ── Compute round-trip cost and minimum required move ──
    # 計算來回成本和最低所需波動
    c_round_pct = compute_round_trip_cost_pct(symbol, volume_24h=volume_24h)
    clamped_wr = max(0.3, min(1.0, win_rate))  # clamp to [0.3, 1.0] / 限制在 [0.3, 1.0]
    min_move_pct = c_round_pct / clamped_wr * 1.3  # 30% safety margin / 30% 安全邊際

    # ── Safety valve: avoid zero-trade days ──
    # 安全閥：避免整天零成交（Operator 決策 2026-04-02）
    # If no trades today and ATR is at least half the round-trip cost, allow entry.
    # This ensures some trades execute even in low-volatility markets.
    # 若今天尚無成交且 ATR 至少為來回成本一半，放行。
    if daily_trade_count == 0 and atr_pct > c_round_pct * 0.5:
        return (False, "daily_safety_valve")

    # ── Core check: ATR must exceed minimum required move ──
    # 核心檢查：ATR 必須超過最低所需波動
    if atr_pct < min_move_pct:
        return (
            True,
            "insufficient_volatility: atr=%.4f < min=%.4f (cost=%.4f wr=%.2f)"
            % (atr_pct, min_move_pct, c_round_pct, clamped_wr),
        )

    return (False, "")

File: rust/openclaw_core/src/cost_gate.rs
//! CostGate — Round-trip cost estimation and cost-based trade rejection
//! CostGate — 往返成本估算及基於成本的交易拒絕
//!
//! MODULE_NOTE (中文):
//!   CostGate 根據 24h 成交量分級估算往返交易成本（滑點 + 手續費），
//!   並判斷 ATR 是否足以覆蓋成本。若預期波動不足以覆蓋成本，則拒絕交易。
//!   安全閥：每日首筆交易且 ATR > 0.5×成本時放行（防止零交易日）。
//!   fail-open：ATR 為 None 時放行（資料不足時不阻塞）。
//!
//! MODULE_NOTE (English):
//!   CostGate estimates round-trip trading costs (slippage + fees) by 24h volume tier,
//!   then checks whether ATR is sufficient to cover costs. Rejects trades when expected
//!   volatility cannot overcome costs.
//!   Safety valve: first trade of day with ATR > 0.5× cost is allowed (prevent zero-trade days).
//!   Fail-open: ATR=None passes (don't block when data is unavailable).

// ---------------------------------------------------------------------------
// Cost tiers by 24h volume / 按 24h 成交量的成本分級
// ---------------------------------------------------------------------------

/// A single cost tier mapping volume threshold to slippage and fee estimates.
/// 單一成本分級：成交量門檻對應滑點和手續費估算。
struct CostTier {
    min_volume_usd: f64,
    slippage_pct: f64,
    taker_fee_pct: f64,
}

/// Volume-based cost tiers (sorted descending by volume).
/// 基於成交量的成本分級（按成交量降序排列）。
const COST_TIERS: &[CostTier] = &[
    CostTier {
        min_volume_usd: 1_000_000_000.0,
        slippage_pct: 0.01,
        taker_fee_pct: 0.055,
    },
    CostTier {
        min_volume_usd: 100_000_000.0,
        slippage_pct: 0.02,
        taker_fee_pct: 0.055,
    },
    CostTier {
        min_volume_usd: 10_000_000.0,
        slippage_pct: 0.05,
        taker_fee_pct: 0.055,
    },
    CostTier {
        min_volume_usd: 1_000_000.0,
        slippage_pct: 0.15,
        taker_fee_pct: 0.055,
    },
    CostTier {
        min_volume_usd: 0.0,
        slippage_pct: 0.30,
        taker_fee_pct: 0.055,
    },
];

/// Compute round-trip cost percentage for a given 24h volume.
/// 計算指定 24h 成交量的往返成本百分比。
///
/// Formula: (taker_fee + slippage) × 2 × 100 (entry + exit).
/// 公式：(taker_fee + slippage) × 2 × 100（入場 + 出場）。
pub fn compute_round_trip_cost_pct(volume_24h: f64) -> f64 {
    let tier = COST_TIERS
        .iter()
        .find(|t| volume_24h >= t.min_volume_usd)
        .unwrap_or(COST_TIERS.last().unwrap());
    (tier.taker_fee_pct + tier.slippage_pct) * 2.0 / 100.0 * 100.0
}

/// Result of the cost gate check / 成本門控檢查結果
pub struct CostGateResult {
    /// Whether the trade should be rejected / 是否應拒絕交易
    pub rejected: bool,
    /// Human-readable reason / 人類可讀原因
    pub reason: String,
}

/// Check whether a trade should be rejected due to insufficient volatility vs cost.
/// 檢查交易是否應因波動性不足以覆蓋成本而被拒絕。
///
/// # Arguments / 參數
/// - `atr_pct`:    ATR as percentage of price (None = fail-open) / ATR 佔價格百分比（None = 放行）
/// - `win_rate`:   Estimated strategy win rate (default 0.5) / 策略預估勝率（預設 0.5）
/// - `daily_trade_count`: Trades executed today so far / 今日已執行交易數
/// - `volume_24h`: 24-hour USD volume for the symbol / 該幣種 24h 美元成交量
pub fn should_reject_for_cost(
    atr_pct: Option<f64>,
    win_rate: f64,
    daily_trade_count: u32,
    volume_24h: f64,
) -> CostGateResult {
    // 1. No ATR data → fail-open (don't block when we lack data)
    //    沒有 ATR 資料 → 放行（資料不足時不阻塞）
    let atr = match atr_pct {
        Some(a) => a,
        None => {
            return CostGateResult {
                rejected: false,
                reason: "no ATR data, fail-open".into(),
            }
        }
    };

    // 2. Compute round-trip cost / 計算往返成本
    let c_round = compute_round_trip_cost_pct(volume_24h);

    // 3. Clamp win rate to [0.3, 1.0] (prevent division issues)
    //    限制勝率在 [0.3, 1.0]（防止除法問題）
    let clamped_wr = win_rate.clamp(0.3, 1.0);

    // 4. Minimum required move = cost / win_rate × safety margin 1.3
    //    最低要求波動 = 成本 / 勝率 × 安全邊際 1.3
    let min_move_pct = c_round / clamped_wr * 1.3;

    // 5. Safety valve: first trade of day AND ATR > 0.5×cost → allow
    //    安全閥：今日首筆交易且 ATR > 0.5×成本 → 放行
    if daily_trade_count == 0 && atr > c_round * 0.5 {
        return CostGateResult {
            rejected: false,
            reason: format!(
                "safety valve: first trade, ATR {:.4}% > 0.5×cost {:.4}%",
                atr,
                c_round * 0.5
            ),
        };
    }

    // 6. Reject if ATR < minimum required move / ATR < 最低要求波動則拒絕
    if atr < min_move_pct {
        CostGateResult {
            rejected: true,
            reason: format!(
                "ATR {:.4}% < min_move {:.4}% (cost={:.4}%, wr={:.2})",
                atr, min_move_pct, c_round, clamped_wr
            ),
        }
    } else {
        CostGateResult {
            rejected: false,
            reason: format!(
                "ATR {:.4}% >= min_move {:.4}% (cost={:.4}%)",
                atr, min_move_pct, c_round
            ),
        }
    }
}

// ===========================================================================
// Tests / 測試
// ===========================================================================
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cost_tier_high_volume() {
        // 1B+ volume → slippage 0.01%, fee 0.055%
        let cost = compute_round_trip_cost_pct(2_000_000_000.0);
        let expected = (0.01 + 0.055) * 2.0;
        assert!((cost - expected).abs() < 1e-9, "high vol cost={cost}");
    }

    #[test]
    fn test_cost_tier_mid_volume() {
        // 50M is >= 10M tier but < 100M → slippage 0.05%
        let cost = compute_round_trip_cost_pct(50_000_000.0);
        let expected = (0.05 + 0.055) * 2.0;
        assert!((cost - expected).abs() < 1e-9, "mid vol cost={cost}");
    }

    #[test]
    fn test_cost_tier_low_volume() {
        // 500K → lowest tier: slippage 0.30%
        let cost = compute_round_trip_cost_pct(500_000.0);
        let expected = (0.30 + 0.055) * 2.0;
        assert!((cost - expected).abs() < 1e-9, "low vol cost={cost}");
    }

    #[test]
    fn test_cost_tier_exact_boundary() {
        // Exactly 10M → should match the 10M tier
        let cost = compute_round_trip_cost_pct(10_000_000.0);
        let expected = (0.05 + 0.055) * 2.0;
        assert!((cost - expected).abs() < 1e-9);
    }

    #[test]
    fn test_reject_no_atr_fail_open() {
        let res = should_reject_for_cost(None, 0.5, 5, 100_000_000.0);
        assert!(!res.rejected, "None ATR should pass: {}", res.reason);
    }

    #[test]
    fn test_reject_low_atr_high_cost() {
        // Low volume (high cost) + tiny ATR → should reject
        let res = should_reject_for_cost(Some(0.1), 0.5, 3, 500_000.0);
        assert!(res.rejected, "Low ATR should be rejected: {}", res.reason);
    }

    #[test]
    fn test_accept_high_atr() {
        // High ATR + high volume (low cost) → should pass
        let res = should_reject_for_cost(Some(2.0), 0.5, 3, 1_000_000_000.0);
        assert!(!res.rejected, "High ATR should pass: {}", res.reason);
    }

    #[test]
    fn test_safety_valve_first_trade() {
        // First trade of day, ATR barely above 0.5×cost → allow
        let cost = compute_round_trip_cost_pct(1_000_000.0);
        let atr = cost * 0.6; // above 0.5×cost but below min_move
        let res = should_reject_for_cost(Some(atr), 0.5, 0, 1_000_000.0);
        assert!(
            !res.rejected,
            "Safety valve should allow first trade: {}",
            res.reason
        );
    }

    #[test]
    fn test_safety_valve_not_first_trade() {
        // Same ATR but not first trade → should reject
        let cost = compute_round_trip_cost_pct(1_000_000.0);
        let atr = cost * 0.6;
        let res = should_reject_for_cost(Some(atr), 0.5, 1, 1_000_000.0);
        assert!(
            res.rejected,
            "Non-first trade with low ATR should reject: {}",
            res.reason
        );
    }

    #[test]
    fn test_win_rate_clamped_low() {
        // Very low win rate (0.1) → clamped to 0.3
        let res1 = should_reject_for_cost(Some(1.5), 0.1, 3, 100_000_000.0);
        let res2 = should_reject_for_cost(Some(1.5), 0.3, 3, 100_000_000.0);
        // Both should yield same result since 0.1 is clamped to 0.3
        assert_eq!(res1.rejected, res2.rejected);
    }

    #[test]
    fn test_zero_volume_uses_lowest_tier() {
        let cost = compute_round_trip_cost_pct(0.0);
        let expected = (0.30 + 0.055) * 2.0;
        assert!((cost - expected).abs() < 1e-9);
    }
}
